Fix/mkdwn-regex

kishore03109 opened the pull request:
## Problem 
This script solves multiple issues with regards to the current state of migrations script. 

Functional: 
1.1. Nested inline markdown 
```
| [![Primary 1 Suggested Reads Image](/images/recommendationsprimary/p1-grl.jpg)](/images/recommendationsprimary/nlb-suggested-reads-primary-p1-fa-ver-300320.pdf) 
```

This issue primarily stems from us not using an off the shelf markdown parser, and this was an edge case that was not caught. As a quick hot fix, the regex now captures nested markdown syntax to cater for this. This was ran against the failing repo `nlb-discovereads` [here](https://github.com/isomerpages/nlb-discovereads/commit/ad656c7b30aa7f32f3d46e5aaba13f58062355be#diff-931eda0023b7643ddf7d79bade5de9ce1e510319e95825115df86015bf666730).

1.2. Fixing encoding issues with yml keys

for the code of 
`favicon: images/hi%20low`

This script did not modify this as it checked if the file existed first. Rather than catering for encoding + different languages which would make this script more complex, I pre-emptively lowercase it, AND then report to eng that the file is missing. This was tested and fixed [here](e76c60a9479e71cc623faa6b97f88b20e943833e).

Dev Facing QOL fixes:

2.1. Lack of awareness of side effects 

This script is NOT idempotent. It seeks to handle e2e migration of an Netlify to Amplify site, and such 
- Creates an amplify site 
- Creates multiple commits to ensure compatibility 
- Create SQL commands 

Eng feedback was that these side effects were not clear. As a result, rather than having a file just to log errors, ALL logs will be appended to the file, with a deliniator of a datetime to signify every run of the script. This way, when the script fails,  end should be more aware of the side effects + should be abale to revert them if needed to rectify. 

![Screenshot 2023-09-07 at 11 16 08 AM](https://github.com/isomerpages/isomer-conversion-scripts/assets/42832651/c964152b-3b0a-4ea6-b3ce-ea4a229ba3ab)


2.2. Lack of awareness that the SQL queries append to the file, it doesn't replace 

While this was intentionally done as a way to store the history of all the sites that has been migrated, this was confusing to dev last week as eng copy-pasted the whole string as-is, leading to unique constraints being hit. As a quick win, every time the script is run, a delineator will be appended. 

This would be similar to the screenshot below (using the same delineator) 
![Screenshot 2023-09-07 at 11 16 08 AM](https://github.com/isomerpages/isomer-conversion-scripts/assets/42832651/5da3bb04-f483-469a-8a5b-98a3be6eff87)

2.3. Failing to load up env vars prior to the script causing eng to rerun 

Rectifying for this script is quite a bit of click-ops. Eg, one would need to go into amplify to delete the created apps if it exists. As a quick win, the script checks for the env var __**first**__ before beginning to do anything.

kishore03109 committed:
fix(script): match nested mkdwn

kishore03109 committed:
fix(script): report migrated sites

kishore03109 committed:
feat(migration script): check for all env vars

kishore03109 committed:
feat(migration script): log all side effects

kishore03109 committed:
style(migration): fix eslint issues

kishore03109 committed:
fix(migration): fx all encoding issues

kishore03109 commented on src/amplify-migration/amplifyMigrationScript.ts:
```diff
@@ -62,6 +58,18 @@ function readCsvFile(
 
 async function main() {
+  // check for all env vars first
+  if (
```
Fix 2.3

kishore03109 commented on src/amplify-migration/amplifyMigrationScript.ts:
```diff
@@ -78,18 +86,37 @@ async function main() {
+
+  // delineate logs for easier separation of runs
+  const delineateString = `------------------${new Date().toLocaleString()}------------------`;
+  fs.appendFileSync(path.join(__dirname, LOGS_FILE), delineateString + os.EOL);
```
fix 2.1

kishore03109 commented on src/amplify-migration/amplifyMigrationScript.ts:
```diff
@@ -78,18 +86,37 @@ async function main() {
+  // delineate logs for easier separation of runs
+  const delineateString = `------------------${new Date().toLocaleString()}------------------`;
+  fs.appendFileSync(path.join(__dirname, LOGS_FILE), delineateString + os.EOL);
+  fs.appendFileSync(
```
fix 2.2

kishore03109 commented on src/amplify-migration/yamlUtils.ts:
```diff
@@ -43,17 +43,19 @@ export async function changeContentInYamlFile(
+  // YAML does not seem to have a way to update the value of a key in place
+  // We just mutate all to lowercase to not care about encoding. Then we report if image is not found
+  // rather than programmatically fixing something that we are not 100% sure of.
+  fileContent = fileContent.replace(oriFilePath, filePath.toLowerCase());
```
fix 1.2

seaerchin replied:
does `setOfAllDocumentsPath` come with lowercased items by default? otherwise when we call the `.has` method, it might be that the set contains the original version but not the lower cased one

kishore03109 replied:
> does setOfAllDocumentsPath come with lowercased items by default?

good qn, and yes, we convert everything to lowercase!

seaerchin replied:
we can consider using a utility [type](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html#lowercasestringtype) here so it's clear to devs + this invariant is maintained.

if this file isn't going to be actively maintained, then it's ok otherwise, it seems important enough (and maybe low effort enough?) to do

kishore03109 replied:
ok have done it [here](https://github.com/isomerpages/isomer-conversion-scripts/pull/70/commits/5b354298a5737236c59b2da35f2cd20ced86be81)

kishore03109 commented on src/amplify-migration/amplifyMigrationScript.ts:
```diff
@@ -350,7 +387,9 @@ export async function changeFileContent({
   }));
 
-  const markdownRegex = /\[(.*?)\]\((.*?)\)/g;
+  // we want to be able to modify nested markdown links
```
fix 1.1

seaerchin commented:
Love the abbreviation HAHA

seaerchin commented on src/amplify-migration/amplifyMigrationScript.ts:
```diff
@@ -350,7 +387,9 @@ export async function changeFileContent({
-  const markdownRegex = /\[(.*?)\]\((.*?)\)/g;
+  // we want to be able to modify nested markdown links
+  // eg. [![inline text](/images/someimage.jpg)](/images/somedoc.pdf)
+  const markdownRegex = /(!?\[([^\]]*)\])?\(([^\s]*)\s*(".*")?\)/g;
```
might want to break this down into the different sub parts then combine into 1 big regex. this makes the code more maintainable + aids clarity. 

i couldn't really make sense of this even after pasting into `regex101` ._. not a priority if we aren't gonna touch the migration script anytime soon

kishore03109 replied:
uhm sorry basic qn, how can i break this down into smaller subparts ah? 
maybe why this became extra complicated is because I am trying to capture this specific case:  ``` [![inline text](/images/someimage.jpg)](/images/somedoc.pdf)``` , where there exists inner + outer links. 

while I could do a 
```
const outerLink = `!?\[([^\]]*)\])?\`
const innerLink = `(([^\s]*)\s*(".*")?\`
const markdownRegex = `/(${outerLink}${outerLink})/g`
```

I am not sure if this is more helpful. Is this what you intended?

seaerchin replied:
i think in this case, we should be using raw strings and combining them using the `RegExp` constructor - see [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions#creating_a_regular_expression); this is in a similar fashion to what you mentioned - if that works, then it's ok too

kishore03109 replied:
7ac78b046ff2a4905b318a66b5497b4fabd51bb2 oki

seaerchin commented on src/amplify-migration/githubUtils.ts:
```diff
@@ -61,3 +61,12 @@ export async function isRepoEmpty(repoName: string): Promise<boolean> {
+export function isRepoMigrated(repoPath: string): boolean {
+  // read the config.yml file
+  const configFile = fs.readFileSync(`${repoPath}/_config.yml`, "utf-8");
+  if (configFile.includes(".amplifyapp.com")) {
+    return true;
+  }
+  return false;
```
nit: can probably return `configFile.includes(".amplifyapp.com")`

also, there were cases previously where the config file didn't update but our db did (ie, our db is the source of truth here). given that this is a script, maybe we wanna log this to the file and ask them to check manually

kishore03109 replied:
@seaerchin The caller of this function already logs if the repo is already migrated! 
also, this does not work if "where the config file didn't update but our db did", because the only moment where this error will come to light is when eng runs the sql commands to which then we will get conflict errors. 
I think this is ok as 
1. errors are not silent 
2. Side effects are clearer now when looking at the logs, eng should be able to remediate this edge case. 


Abit of context here is that this change was introduced because there exists human error when ops gives us the list (which should be on us to encounter for)

seaerchin replied:
yeap, for clarity's sake, what i meant is that we should be logging to the file (or in the caller) when the config file does **not** include the `amplifyapp` post migration. 

this would give more salient feedback faster also

kishore03109 replied:
> yeap, for clarity's sake, what i meant is that we should be logging to the file (or in the caller) when the config file does not include the amplifyapp post migration.

hey @seaerchin I hear you on this, but I realised that this is quite easily solvable! Problem stems when the site migrated is an older site that does not have the `staging` + `prod` key. Rather than logging + making eng add in the two lines, opting for the quick win [here](https://github.com/isomerpages/isomer-conversion-scripts/pull/70/commits/5c472d874034528a212531f53ec5faecbc1ea40d) of just appending the key value pair directly onto the `config.yml` file to reduce the mental overhead required.

seaerchin commented on src/amplify-migration/jsDomUtils.ts:
skimmed, seems like formatting changes

kishore03109 replied:
yes, this is the case

kishore03109 committed:
style(githubUtils): cleaner func

kishore03109 committed:
style(migration script): change file to save

alexanderleegs reviewed (commented):
Just checking, does this PR also fix the issue for links which are not prepended with `/` (because netlify converts these to relative paths, unlike amplify), or are we planning to do that in a separate PR?

kishore03109 committed:
fix(amp script): fix parser crashing when complex html is used

QOL improvements: Logging + adding of prod/stging keys if dont exist +
check remote for migration was already done

harishv7 committed:
fix: README and package json

kishore03109 committed:
refactor(script): make regex more readable

kishore03109 commented:
@alexanderleegs  done in another PR please

kishore03109 committed:
chore(readme): inform devs about

dcshzj commented on src/amplify-migration/amplifyMigrationScript.ts:
```diff
@@ -350,7 +391,12 @@ export async function changeFileContent({
+  // eg. [![inline text](/images/someimage.jpg)](/images/somedoc.pdf)
+  // TODO: take care of edge case: [![blah](/images/blah.jpg){:style="width: 300px"}](https://blah.com)
+  const outerLink = `(!?\\[([^\\]]*)\\])`;
+  const innerLink = `\\(([^\\s]*)\\s*(".*")?\\)`;
```
I might not be understanding the regex here, should the quotation marks be replaced with the `[` and `]` brackets instead?

kishore03109 replied:
Hey we can chat about this later in the office, but the quotation marks here refer to matching the characters directly. 

Here are some trivial test cases for this when using an arbitrary regex [tester](https://regex101.com/):
```
[![inline text](/images/someimage.jpg)](/images/somedoc.pdf)
[blah](images/someimage.jpg)
[blah](/images/someimage.jpg)
```
<img width="989" alt="Screenshot 2023-10-09 at 9 08 33 AM" src="https://github.com/isomerpages/isomer-conversion-scripts/assets/42832651/554ec500-9dd9-4f87-bfee-d06f883c5323">
